# Preprocessing

In [79]:
import pandas as pd
import numpy as np

In [80]:
from functions import *

### Import the datasets

In [81]:
# Open the raw datasets
# Heavy Metal
hm_df = pd.read_excel('data/AllHMFiles.xlsx', engine='openpyxl').fillna(0.0)

# Pesticides
pesticides_df = pd.read_excel('data/AllPesticidesFiles.xlsx', engine='openpyxl').fillna(0.0)
classe_pesticides_df = pd.read_excel('data/pesticides.xlsx', engine='openpyxl')

# Surfaces
surfaces_df = pd.read_excel('data/absSurfs.xlsx', engine='openpyxl')

# Distances
distances_df = pd.read_excel('data/distsOneSheet.xlsx', engine='openpyxl')

In [82]:
hm_df.head(5)

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  
0     0.032    40     0.0  
1     0.029    33     0.0  
2     0.038    37     0.0  
3     0.031    37     0.0  
4     0.052    35     0.0

In [83]:
pesticides_df.head(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Nicosulfuron  \
0           0.0           0.0          0.0        0.0  ...           0.0   
1           0.0           0.0          0.0        0.0  ...           0.0   
2           0.0           0.0          0.0        0.0  ...           0.0   
3           0.0           0.0          0.0        0.0  ...           0.0   
4           0.0           0.0          0.0        0.0  ...           0.0   

   Piperonylbutoxide  Prosulfuron  Simazine  Terbuthylazine  Dieldrin  \
0                0.0          0.0       0.0             0.0       0.0   
1                0.0          0.0       0.0             0.0       0.0   
2                0.0          0.0       0.0             0.0       0.0   
3                0.0          0.0       0.0             0.0       0.0   
4                0.0          0.0       0.0             0.0       0.0   

   Fenamidone  Trifluralin  Cycloate  Pentachloroanisol  
0         0.0          0.0       0.0                0.0  
1         0.0          0.0       0.0                0.0  
2         0.0          0.0       0.0                0.0  
3         0.0          0.0       0.0                0.0  
4         0.0          0.0       0.0                0.0  

[5 rows x 117 columns]

In [84]:
classe_pesticides_df.head(5)

importName           name    statusEN   LMR  \
0      Aclonifen      aclonifen  authorized  0.05   
1         Aldrin         aldrin  prohibited  0.01   
2   Ametoctradin   ametoctradin  authorized  0.05   
3  Anthraquinone  anthraquinone  prohibited  0.02   
4   Azoxystrobin   azoxystrobin  authorized  0.05   

                          familyEN          typeEN  
0                   diphenyl ether       herbicide  
1                   organochlorine     insecticide  
2               triazolopyrimidine       fungicide  
3  polycyclic aromatic hydrocarbon  bird repellant  
4                      strobilurin       fungicide

In [85]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [86]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

### Heavy Metal

In [87]:
hm_df.head()

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  
0     0.032    40     0.0  
1     0.029    33     0.0  
2     0.038    37     0.0  
3     0.031    37     0.0  
4     0.052    35     0.0

In [88]:
MRL_HM_dict = {
    'Lead' : [1.5],
    'Cadmium' : [1],
    'Arsenic' : [0.3],
    'Mercury' : [1],
    'Copper' : [30]
}

MRL_HM = pd.DataFrame.from_dict(MRL_HM_dict)
MRL_HM

Lead  Cadmium  Arsenic  Mercury  Copper
0   1.5        1      0.3        1      30

In [89]:
hm_df['mrl_lead'] = np.where(hm_df['Lead'] > 1.5, 1, 0)
hm_df['mrl_cadmium'] = np.where(hm_df['Cadmium'] > 1, 1, 0)
hm_df['mrl_arsenic'] = np.where(hm_df['Arsenic'] > 0.3, 1, 0)
hm_df['mrl_mercury'] = np.where(hm_df['Mercury'] > 1, 1, 0)
hm_df['mrl_copper'] = np.where(hm_df['Copper'] > 30, 1, 0)

hm_df.head()

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_mercury  \
0     0.032    40     0.0         0            0            0            0   
1     0.029    33     0.0         0            0            0            0   
2     0.038    37     0.0         0            0            0            0   
3     0.031    37     0.0         0            0            0            0   
4     0.052    35     0.0         0            0            0            0   

   mrl_copper  
0           0  
1           0  
2           0  
3           0  
4           0

In [90]:
# Dropping columns of metals without specific threshhold level defined by laws
hm_df = hm_df.drop(['Chromium'], axis=1)
hm_df = hm_df.drop(['Chrome'], axis=1)
hm_df = hm_df.drop(['Zinc'], axis=1)

# Dropping other columns not needed anymore
hm_df = hm_df.drop(['Lead'], axis=1)
hm_df = hm_df.drop(['Cadmium'], axis=1)
hm_df = hm_df.drop(['Arsenic'], axis=1)
hm_df = hm_df.drop(['Mercury'], axis=1)
hm_df = hm_df.drop(['Copper'], axis=1)

hm_df.head()

REF....SUBSTANCE   PERIOD  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_mercury  \
0            B0017  2017/P1         0            0            0            0   
1            B0019  2017/P1         0            0            0            0   
2            B0021  2017/P1         0            0            0            0   
3            B0023  2017/P1         0            0            0            0   
4            B0037  2017/P1         0            0            0            0   

   mrl_copper  
0           0  
1           0  
2           0  
3           0  
4           0

In [91]:
group_hm_df = hm_df.groupby('REF....SUBSTANCE').agg(max)
group_hm_df.head()

PERIOD  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_mercury  \
REF....SUBSTANCE                                                             
B0012             2020/P4         0            0            0            0   
B0017             2020/P4         0            0            0            0   
B0019             2020/P4         0            0            0            0   
B0021             2020/P4         0            1            0            0   
B0023             2020/P4         0            1            0            0   

                  mrl_copper  
REF....SUBSTANCE              
B0012                      0  
B0017                      0  
B0019                      0  
B0021                      0  
B0023                      0

### Pesticides

Fungicide :

In [14]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].head(5)

importName                       name    statusEN   LMR  \
2         Ametoctradin               ametoctradin  authorized  0.05   
4         Azoxystrobin               azoxystrobin  authorized  0.05   
5            Benalaxyl                  benalaxyl  prohibited  0.05   
7  Benthiavalicarb.iso  benthiavalicarb-isopropyl  authorized  0.05   
8     Benzovindiflupyr           benzovindiflupyr  authorized  0.05   

             familyEN     typeEN  
2  triazolopyrimidine  fungicide  
4         strobilurin  fungicide  
5        phenylamides  fungicide  
7           carbamate  fungicide  
8               amide  fungicide

In [15]:
fungicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].importName.to_list()
fungicides.append('Diphenylamine')
fungicides.append('Tebuconazole')
fungicides

['Ametoctradin',
 'Azoxystrobin',
 'Benalaxyl',
 'Benthiavalicarb.iso',
 'Benzovindiflupyr',
 'Boscalid',
 'Bupirimate',
 'Captan',
 'Carbendazim',
 'Chlorothalonil',
 'Cyazofamid',
 'Cyflufenamid',
 'Cymoxanil',
 'Cyprodinil',
 'Difenoconazole',
 'Dimethomorph',
 'Dimoxystrobin',
 'Dodine',
 'Epoxiconazole',
 'Ethirimol',
 'Famoxadone',
 'Fenamidone',
 'Fenbuconazole',
 'Fenhexamid',
 'Fenpropidin',
 'Fenpropimorph',
 'Fluazinam',
 'Fludioxonil',
 'Fluopicolide',
 'Fluopyram',
 'Flutolanil',
 'Fluxapyroxad',
 'Hexaconazole',
 'Iprovalicarb',
 'Isofetamid',
 'Mandipropamid',
 'Metalaxyl',
 'Metconazole',
 'Metrafenone',
 'Penconazole',
 'Pencycuron',
 'Pentachloroanisol',
 'Prochloraz',
 'Propamocarb',
 'Propiconazole',
 'Prothioconazole_d',
 'Pyraclostrobin',
 'Pyrimethanil',
 'Quinoxyfen',
 'Spiroxamine',
 'Tetraconazole',
 'Thiabendazole',
 'Thiophanate.methyl',
 'Trifloxystrobin',
 'X2.phenylphenol',
 'Zoxamide',
 'Diphenylamine',
 'Tebuconazole']

Insecticide :

In [16]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].head(5)

importName                 name          statusEN   LMR  \
1                Aldrin               aldrin        prohibited  0.01   
13  Chlorantraniliprole  chlorantraniliprole        authorized  0.05   
17  Chlorpyrifos.methyl  chlorpyrifos-methyl        prohibited  0.05   
19            Coumaphos            coumaphos        prohibited  0.10   
20         Cyanofenphos         cyanofenphos  not yet assessed   NaN   

               familyEN       typeEN  
1        organochlorine  insecticide  
13  anthranilic diamide  insecticide  
17      organophosphate  insecticide  
19        thiophosphate  insecticide  
20      organophosphate  insecticide

In [17]:
insecticides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].importName.to_list()
insecticides.append('Diphenylamine')
insecticides.append('Methiocarb_total')
insecticides

['Aldrin',
 'Chlorantraniliprole',
 'Chlorpyrifos.methyl',
 'Coumaphos',
 'Cyanofenphos',
 'Cyflumetofen',
 'Cypermethrin',
 'Deet',
 'Dieldrin',
 'Etofenprox',
 'Fenoxycarb',
 'Fenpropathrin',
 'Flonicamid',
 'Flucythrinate',
 'Flupyradifurone',
 'Imidacloprid',
 'Indoxacarb',
 'Methoxyfenozide',
 'Phosmet',
 'Piperonylbutoxide',
 'Pirimicarb',
 'Spinosad',
 'Spirotetramat',
 'Tau.fluvalinate',
 'Tebufenozide',
 'Tetradifon',
 'Thiacloprid',
 'Diphenylamine',
 'Methiocarb_total']

Herbicide : 

In [18]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].head(5)

importName           name    statusEN   LMR          familyEN  \
0       Aclonifen      aclonifen  authorized  0.05    diphenyl ether   
6     Benfluralin    benfluralin  authorized  0.05    dinitroaniline   
15  Chlorotoluron  chlorotoluron  authorized  0.05        phenylurea   
16   Chlorpropham   chlorpropham  prohibited  0.05         carbamate   
18      Clethodim      clethodim  authorized  0.05  cyclohexanedione   

       typeEN  
0   herbicide  
6   herbicide  
15  herbicide  
16  herbicide  
18  herbicide

In [19]:
herbicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].importName.to_list()
herbicides

['Aclonifen',
 'Benfluralin',
 'Chlorotoluron',
 'Chlorpropham',
 'Clethodim',
 'Cycloate',
 'Diflufenican',
 'Ethofumesate',
 'Fluazifop',
 'Flufenacet',
 'Metamitron',
 'Methabenzthiazuron',
 'Metobromuron',
 'Metolachlor',
 'Nicosulfuron',
 'Pendimethalin',
 'Phenmedipham',
 'Propanil',
 'Propyzamide',
 'Prosulfocarb',
 'Prosulfuron',
 'Simazine',
 'Terbuthylazine',
 'Tri.allate',
 'Trifluralin']

Others :

In [20]:
classe_pesticides_df[(classe_pesticides_df['typeEN'] != 'fungicide') & 
                     (classe_pesticides_df['typeEN'] != 'insecticide') & 
                     (classe_pesticides_df['typeEN'] != 'herbicide')]

importName           name    statusEN   LMR  \
3           Anthraquinone  anthraquinone  prohibited  0.02   
34          Diphenylamine  diphenylamine  prohibited  0.05   
69       Methiocarb_total     methiocarb  prohibited  0.05   
86             Propargite     propargite  prohibited  0.05   
100          Tebuconazole   tebuconazole  authorized  0.05   
112  X6.benzylaminopurine  benzyladenine  authorized   NaN   
114                   NaN            NaN         NaN   NaN   

                            familyEN  \
3    polycyclic aromatic hydrocarbon   
34                             amine   
69                         carbamate   
86                               NaN   
100                         triazole   
112                       cytokinins   
114                              NaN   

                                           typeEN  
3                                  bird repellant  
34    growth regulator, fungicide and insecticide  
69   insecticide, molluscicide and bird repellent  
86                                      acaricide  
100                fungicide and growth regulator  
112                              growth regulator  
114                                           NaN

##### Choice to treat the others

Adding to the corresponding categories, stay in "other" ? What about the growth regulator ?

#### Add some columns

In [21]:
pesticides_df['qtt_fungicides'] = pesticides_df.apply(qtt_pesticides, args = fungicides, axis = 1)
pesticides_df['qtt_insecticides'] = pesticides_df.apply(qtt_pesticides, args = insecticides, axis = 1)
pesticides_df['qtt_herbicides'] = pesticides_df.apply(qtt_pesticides, args = herbicides, axis = 1)

In [22]:
pesticides_df['nb_fungicides'] = pesticides_df.apply(nb_pesticides, args = fungicides, axis = 1)
pesticides_df['nb_insecticides'] = pesticides_df.apply(nb_pesticides, args = insecticides, axis = 1)
pesticides_df['nb_herbicides'] = pesticides_df.apply(nb_pesticides, args = herbicides, axis = 1)

In [23]:
pesticides_df['lmr_fungicides'] = pesticides_df.apply(lmr_pesticides, args = fungicides, axis = 1)
pesticides_df['lmr_insecticides'] = pesticides_df.apply(lmr_pesticides, args = insecticides, axis = 1)
pesticides_df['lmr_herbicides'] = pesticides_df.apply(lmr_pesticides, args = herbicides, axis = 1)

In [24]:
pesticides_df.head(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   
5            B0039  2017/P1     0.0     0.0000          0.00           0.0   
6            B0041  2017/P1     0.0     0.0000          0.00           0.0   
7            B0043  2017/P1     0.0     0.0000          0.00           0.0   
8            B0045  2017/P1     0.0     0.0000          0.00           0.0   
9            B0059  2017/P1     0.0     0.0000          0.00           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Pentachloroanisol  \
0          0.00           0.0         0.00        0.0  ...                0.0   
1          0.00           0.0         0.00        0.0  ...                0.0   
2          0.00           0.0         0.00        0.0  ...                0.0   
3          0.00           0.0         0.00        0.0  ...                0.0   
4          0.00           0.0         0.00        0.0  ...                0.0   
5          0.01           0.0         0.00        0.0  ...                0.0   
6          0.00           0.0         0.00        0.0  ...                0.0   
7          0.00           0.0         0.01        0.0  ...                0.0   
8          0.00           0.0         0.00        0.0  ...                0.0   
9          0.00           0.0         0.00        0.0  ...                0.0   

   qtt_fungicides  qtt_insecticides  qtt_herbicides  nb_fungicides  \
0          0.0000            0.0000            0.00              0   
1          0.0188            0.0000            0.00              1   
2          0.0711            0.0000            0.00              1   
3          0.0294            0.0000            0.00              1   
4          0.0000            0.0000            0.01              0   
5          0.0000            0.0208            0.00              0   
6          0.0000            0.0000            0.00              0   
7          0.0100            0.0000            0.00              1   
8          0.0000            0.0000            0.00              0   
9          0.0145            0.0000            0.00              1   

   nb_insecticides  nb_herbicides  lmr_fungicides  lmr_insecticides  \
0                0              0               0                 0   
1                0              0               0                 0   
2                0              0               1                 0   
3                0              0               0                 0   
4                0              1               0                 0   
5                2              0               0                 0   
6                0              0               0                 0   
7                0              0               0                 0   
8                0              0               0                 0   
9                0              0               0                 0   

   lmr_herbicides  
0               0  
1               0  
2               0  
3               0  
4               0  
5               0  
6               0  
7               0  
8               0  
9               0  

[10 rows x 126 columns]

In [25]:
pesticides_df['year'] = pesticides_df['PERIOD'].apply(lambda x: x[:4])
pesticides_df['trimester'] = pesticides_df['PERIOD'].apply(lambda x: x[-1:])
pesticides_df.sample(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
219            F0011  2019/P1     0.0     0.0000           0.0           0.0   
370            F0021  2020/P1     0.0     0.0000           0.0           0.0   
154            B0061  2018/P3     0.0     0.0097           0.0           0.0   
261            F0011  2019/P2     0.0     0.0000           0.0           0.0   
203            B0043  2019/P1     0.0     0.0000           0.0           0.0   
471            F0033  2020/P3     0.0     0.0000           0.0           0.0   
65             F0011  2017/P3     0.0     0.0000           0.0           0.0   
91             B0103  2018/P1     0.0     0.0000           0.0           0.0   
337            B0012  2020/P1     0.0     0.0000           0.0           0.0   
221            F0015  2019/P1     0.0     0.0000           0.0           0.0   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
219           0.0           0.0          0.0        0.0  ...   
370           0.0           0.0          0.0        0.0  ...   
154           0.0           0.0          0.0        0.0  ...   
261           0.0           0.0          0.0        0.0  ...   
203           0.0           0.0          0.0        0.0  ...   
471           0.0           0.0          0.0        0.0  ...   
65            0.0           0.0          0.0        0.0  ...   
91            0.0           0.0          0.0        0.0  ...   
337           0.0           0.0          0.0        0.0  ...   
221           0.0           0.0          0.0        0.0  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
219               0.0          0.0000              0                0   
370               0.0          0.0177              9                0   
154               0.0          0.0000              1                0   
261               0.0          0.0000              0                0   
203               0.0          0.0000              1                0   
471               0.0          0.0000              0                0   
65                0.0          0.0000              0                0   
91                0.0          0.0000              0                0   
337               0.0          0.0048              1                0   
221               0.0          0.0000              0                0   

     nb_herbicides  lmr_fungicides  lmr_insecticides  lmr_herbicides  year  \
219              0               0                 0               0  2019   
370              1               1                 0               0  2020   
154              0               0                 0               0  2018   
261              0               0                 0               0  2019   
203              0               0                 0               0  2019   
471              0               0                 0               0  2020   
65               0               0                 0               0  2017   
91               0               0                 0               0  2018   
337              1               1                 0               0  2020   
221              0               0                 0               0  2019   

     trimester  
219          1  
370          1  
154          3  
261          2  
203          1  
471          3  
65           3  
91           1  
337          1  
221          1  

[10 rows x 128 columns]

##### Alternate (smaller) dataset for pesticides

In [26]:
pesti_names = pesticides_df.columns.to_list()[1:-11]

In [27]:
small_pesticides_df = pesticides_df.drop(pesti_names, axis=1)
small_pesticides_df.head(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
0            B0017          0.0000               0.0            0.00   
1            B0019          0.0188               0.0            0.00   
2            B0021          0.0711               0.0            0.00   
3            B0023          0.0294               0.0            0.00   
4            B0037          0.0000               0.0            0.01   

   nb_fungicides  nb_insecticides  nb_herbicides  lmr_fungicides  \
0              0                0              0               0   
1              1                0              0               0   
2              1                0              0               1   
3              1                0              0               0   
4              0                0              1               0   

   lmr_insecticides  lmr_herbicides  year trimester  
0                 0               0  2017         1  
1                 0               0  2017         1  
2                 0               0  2017         1  
3                 0               0  2017         1  
4                 0               0  2017         1

##### Alternate (group by) dataset for pesticides

In [28]:
group_pesti_df = pesticides_df.groupby('REF....SUBSTANCE').agg(max)

In [29]:
group_pesti_df = group_pesti_df.loc[:,'lmr_fungicides' : 'lmr_herbicides']
group_pesti_df

lmr_fungicides  lmr_insecticides  lmr_herbicides
REF....SUBSTANCE                                                  
B0012                          1                 0               0
B0017                          1                 0               0
B0019                          0                 0               0
B0021                          1                 0               0
B0023                          0                 0               0
B0032                          1                 0               0
B0037                          1                 1               0
B0039                          1                 0               0
B0041                          1                 0               0
B0043                          1                 0               1
B0045                          1                 0               0
B0059                          1                 0               0
B0061                          0                 0               0
B0067                          1                 0               0
B0069                          1                 1               0
B0077                          1                 1               0
B0093                          1                 0               0
B0096                          0                 0               0
B0103                          1                 0               0
B0105                          1                 1               1
B0116                          0                 0               0
B0128                          0                 0               0
B0140                          0                 0               0
B0142                          1                 0               0
B0152                          1                 0               0
B0158                          1                 0               0
B0163                          1                 0               0
F0001                          1                 1               0
F0003                          1                 0               0
F0005                          0                 0               0
F0007                          0                 0               0
F0009                          0                 0               0
F0011                          0                 1               0
F0013                          0                 0               0
F0015                          1                 0               0
F0017                          0                 0               0
F0019                          1                 0               0
F0020                          1                 0               0
F0021                          1                 0               0
F0023                          0                 0               1
F0025                          0                 0               0
F0029                          1                 0               0
F0031                          1                 1               0
F0033                          0                 0               0
F0035                          0                 0               1
F0039                          1                 0               0
F0041                          0                 0               0
F0043                          0                 0               0
F0055                          1                 0               0
F0057                          1                 0               0
F0061                          1                 0               0
F0063                          1                 0               0
F0065                          0                 0               0
F0069                          0                 0               1
F0071                          1                 0               0
F0073                          1                 1               0
FE0001                         0                 0               0

### Sites

##### Getting the area for each zone

In [30]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

In [31]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [32]:
# surfaces_df['country'] = surfaces_df['Site'].apply(lambda x: 'FR' if x.startswith('F') else 'BE')
# surfaces_df['country'].value_counts()

In [33]:
sites_df = distances_df

In [34]:
sites_df['surface'] = sites_df.apply(get_surface, axis = 1)

In [35]:
sites_df.sample(10)

Site     polyID  classCLC         dist       surface
462  F0039   FR-93647       231   809.745662  4.436061e+05
502  F0061  FR-205929       311     0.000000  2.484938e+06
236  B0043    BE-8297       211     0.000000  2.969686e+06
440  B0105   BE-13897       242   679.531173  3.079806e+05
498  F0061   FR-18801       112    14.450700  2.735732e+06
512  F0063   FR-19084       112  1006.085409  3.419785e+05
316  F0017   FR-48732       211  1265.434621  9.580098e+05
0    F0069   FR-88090       231  1183.022362  8.899784e+05
584  F0011  FR-249671       322   517.651747  1.523829e+06
234  B0043    BE-3868       112  1404.705569  3.943985e+05

In [36]:
sites_df['country'] = sites_df['polyID'].apply(lambda x: 'FR' if x.startswith('FR') else 'BE')
sites_df.sample(10)

Site     polyID  classCLC         dist       surface country
281  B0059   BE-24836       511   152.035270  4.582974e+05      BE
148  F0043   FR-19630       112   680.832265  1.300245e+06      FR
86   B0142     BE-857       121  1074.953441  3.229956e+06      BE
449  B0103   BE-18137       242  1422.246533  2.385882e+06      BE
590  F0007  FR-158659       243  1268.118767  9.388673e+04      FR
144  F0041   FR-33583       141   698.719633  3.536624e+05      FR
533  F0013  FR-261806       324  1210.896703  1.081138e+05      FR
62   F0073  FR-214982       312   179.948760  4.420773e+06      FR
427  B0077    BE-4487       122  1339.694031  1.175557e+05      BE
256  F0035   FR-71410       231   207.968262  1.563096e+06      FR

In [37]:
sites_df['country'].value_counts()

FR    335
BE    297
Name: country, dtype: int64

In [38]:
sites_df = sites_df.drop(['polyID'], axis=1)
sites_df.sample(5)

Site  classCLC         dist       surface country
22   F0027       211     0.000000  3.221110e+06      FR
306  F0055       324   444.674516  9.163058e+05      FR
272  B0059       242  1054.380546  1.382229e+06      BE
217  B0041       231  1362.336434  4.150804e+05      BE
206  B0045       511  1142.543367  2.213737e+05      BE

##### Get one dataset with : beehives and area/distance for each zone

In [39]:
surfaces_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [40]:
classCLC = surfaces_df.columns.to_list()[1:]

In [41]:
by_sites_df = surfaces_df.copy(deep=True)
by_sites_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [42]:
for CLC in classCLC:
    by_sites_df[CLC] = by_sites_df.apply(distance_area, args = (CLC,), axis = 1)

In [43]:
by_sites_df

Site  111           112           121         122  123  124  131  132  \
0   F0069  0.0  1.442780e+03  3.587087e+05    0.000000  0.0  0.0  0.0  0.0   
1   F0025  0.0  0.000000e+00  0.000000e+00  700.801763  0.0  0.0  0.0  0.0   
2   F0027  0.0  2.604161e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
3   B0128  0.0  2.412962e+03  4.053755e+06  523.436355  0.0  0.0  0.0  0.0   
4   F0023  0.0  0.000000e+00  7.907522e+02    0.000000  0.0  0.0  0.0  0.0   
..    ...  ...           ...           ...         ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
59  B0069  0.0  7.328884e+03  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
60  B0152  0.0  4.575279e+06  5.227880e+02  421.402885  0.0  0.0  0.0  0.0   
61  F0003  0.0  3.859675e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
62  B0158  0.0  1.574508e+02  0.000000e+00  392.706036  0.0  0.0  0.0  0.0   

    133  ...  411  412  421  422  423  511         512  521  522  523  
0   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
1   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
2   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
3   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
4   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  463.344615  0.0  0.0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...         ...  ...  ...  ...  
58  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
59  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
60  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
61  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
62  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  

[63 rows x 45 columns]

check if correct

In [44]:
distance = distances_df[(distances_df['Site'] == 'F0023') & 
                        (distances_df['classCLC'] == 512)]['dist'].values
distance

array([614.50674791])

In [45]:
area = surfaces_df[surfaces_df['Site'] == 'F0023']['512'].values[0]
area

284728.392578125

In [46]:
area/distance[0]

463.3446150839585

### Merging the datasets into two

3 datasets to 2 : sites + (heavy metal || pesticides)

##### Heavy Metal

##### Pesticides

full

In [47]:
pesticides_df.sample(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
49             B0023  2017/P3     0.0     0.0000           0.0           0.0   
285            B0043  2019/P3     0.0     0.0000           0.0           0.0   
318            B0069  2019/P4     0.0     0.0086           0.0           0.0   
210            B0105  2019/P1     0.0     0.0000           0.0           0.0   
168            F0015  2018/P3     0.0     0.0285           0.0           0.0   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
49            0.0           0.0        0.000     0.0000  ...   
285           0.0           0.0        0.000     0.0000  ...   
318           0.0           0.0        0.417     0.2741  ...   
210           0.0           0.0        0.000     0.0105  ...   
168           0.0           0.0        0.000     0.0000  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
49              0.000          0.0000              0                0   
285             0.000          0.0000              2                0   
318             0.019          0.0000             14                1   
210             0.000          0.0503              3                0   
168             0.000          0.0000              1                0   

     nb_herbicides  lmr_fungicides  lmr_insecticides  lmr_herbicides  year  \
49               0               0                 0               0  2017   
285              0               0                 0               0  2019   
318              0               1                 0               0  2019   
210              2               0                 0               0  2019   
168              0               0                 0               0  2018   

     trimester  
49           3  
285          3  
318          4  
210          1  
168          3  

[5 rows x 128 columns]

In [48]:
small_pesticides_df.sample(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
307            F0041          0.0089            0.0204          0.0000   
335            F0043          0.0050            0.0000          0.0000   
359            B0163          0.1390            0.0000          0.0186   
84             B0043          0.0000            0.0000          0.0139   
366            F0015          0.1279            0.0000          0.0000   

     nb_fungicides  nb_insecticides  nb_herbicides  lmr_fungicides  \
307              2                1              0               0   
335              1                0              0               0   
359              2                0              1               1   
84               0                0              1               0   
366              1                0              0               1   

     lmr_insecticides  lmr_herbicides  year trimester  
307                 0               0  2019         3  
335                 0               0  2019         4  
359                 0               0  2020         1  
84                  0               0  2018         1  
366                 0               0  2020         1

In [49]:
group_pesti_df.sample(5)

lmr_fungicides  lmr_insecticides  lmr_herbicides
REF....SUBSTANCE                                                  
F0009                          0                 0               0
B0041                          1                 0               0
B0128                          0                 0               0
F0055                          1                 0               0
F0019                          1                 0               0

##### sites

In [50]:
sites_df.sample(5)

Site  classCLC         dist       surface country
614  F0003       211  1182.440510  1.869143e+05      FR
545  F0015       311   102.771173  5.107448e+06      FR
502  F0061       311     0.000000  2.484938e+06      FR
381  B0096       243  1484.640512  9.364097e+05      BE
398  B0017       242  1158.849334  4.197580e+05      BE

In [51]:
by_sites_df.sample(5)

Site  111           112           121         122  123  124  131  132  \
3   B0128  0.0  2.412962e+03  4.053755e+06  523.436355  0.0  0.0  0.0  0.0   
56  F0009  0.0  0.000000e+00  0.000000e+00   65.325567  0.0  0.0  0.0  0.0   
38  B0023  0.0  0.000000e+00  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
40  B0021  0.0  0.000000e+00  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
37  B0096  0.0  5.206697e+06  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   

    133  ...  411           412  421  422  423  511  512  521  522  523  
3   0.0  ...  0.0  0.000000e+00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
56  0.0  ...  0.0  0.000000e+00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
38  0.0  ...  0.0  1.809062e+06  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
40  0.0  ...  0.0  2.574868e+03  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
37  0.0  ...  0.0  0.000000e+00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 45 columns]

#### Merge

In [52]:
group_pesti_df.index.name

'REF....SUBSTANCE'

In [53]:
group_pesti_df = group_pesti_df.reset_index()

In [54]:
group_pesti_df = group_pesti_df.rename(columns={"REF....SUBSTANCE": "Site"})

In [55]:
group_pesti_df.head(3)

Site  lmr_fungicides  lmr_insecticides  lmr_herbicides
0  B0012               1                 0               0
1  B0017               1                 0               0
2  B0019               0                 0               0

In [56]:
by_sites_df.head(3)

Site  111          112            121         122  123  124  131  132  \
0  F0069  0.0  1442.780424  358708.730957    0.000000  0.0  0.0  0.0  0.0   
1  F0025  0.0     0.000000       0.000000  700.801763  0.0  0.0  0.0  0.0   
2  F0027  0.0   260.416070       0.000000    0.000000  0.0  0.0  0.0  0.0   

   133  ...  411  412  421  422  423  511  512  521  522  523  
0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[3 rows x 45 columns]

In [57]:
merged_dataset = pd.merge(group_pesti_df, by_sites_df, on=["Site"])

In [58]:
merged_dataset

Site  lmr_fungicides  lmr_insecticides  lmr_herbicides           111  \
0    B0012               1                 0               0    209.227976   
1    B0017               1                 0               0      0.000000   
2    B0019               0                 0               0      0.000000   
3    B0021               1                 0               0      0.000000   
4    B0023               0                 0               0      0.000000   
5    B0032               1                 0               0      0.000000   
6    B0037               1                 1               0      0.000000   
7    B0039               1                 0               0   3098.057557   
8    B0041               1                 0               0    391.077976   
9    B0043               1                 0               1      0.000000   
10   B0045               1                 0               0      0.000000   
11   B0059               1                 0               0      0.000000   
12   B0061               0                 0               0      0.000000   
13   B0067               1                 0               0      0.000000   
14   B0069               1                 1               0      0.000000   
15   B0077               1                 1               0      0.000000   
16   B0093               1                 0               0      0.000000   
17   B0096               0                 0               0      0.000000   
18   B0103               1                 0               0      0.000000   
19   B0105               1                 1               1      0.000000   
20   B0116               0                 0               0      0.000000   
21   B0128               0                 0               0      0.000000   
22   B0140               0                 0               0   1416.095618   
23   B0142               1                 0               0      0.000000   
24   B0152               1                 0               0      0.000000   
25   B0158               1                 0               0      0.000000   
26   B0163               1                 0               0  15107.362039   
27   F0001               1                 1               0      0.000000   
28   F0003               1                 0               0      0.000000   
29   F0005               0                 0               0      0.000000   
30   F0007               0                 0               0      0.000000   
31   F0009               0                 0               0      0.000000   
32   F0011               0                 1               0      0.000000   
33   F0013               0                 0               0      0.000000   
34   F0015               1                 0               0      0.000000   
35   F0017               0                 0               0      0.000000   
36   F0019               1                 0               0      0.000000   
37   F0020               1                 0               0      0.000000   
38   F0021               1                 0               0      0.000000   
39   F0023               0                 0               1      0.000000   
40   F0025               0                 0               0      0.000000   
41   F0029               1                 0               0      0.000000   
42   F0031               1                 1               0      0.000000   
43   F0033               0                 0               0      0.000000   
44   F0035               0                 0               1      0.000000   
45   F0039               1                 0               0      0.000000   
46   F0041               0                 0               0      0.000000   
47   F0043               0                 0               0      0.000000   
48   F0055               1                 0               0      0.000000   
49   F0057               1                 0               0      0.000000   
50   F0061               1  

In [59]:
merged_dataset.to_csv('merged_dataset.csv')

### Changing the NA

# Questions to ask

- When NaN => information irrelevant or not analysed?
- Difference France/Belgium?
- Is the fact that an insecticide is unauthorized relevant? (=> not the same law in France/Belgium)
- Or do we have to show that even banned pesticides are still found?
- How do we treat other categorical pesticides?
- How to increase the number of datas?

### Bo-bo-bonus

- FR  :  36 beehives for 335 zones
- BE  :  27 beehives for 297 zones

Here the CLC zone are left separate, could group some to test